lib/charms/traefik_k8s/v2/ingress.py

```diff
@@ -646,9 +663,22 @@ def provide_ingress_requirements(self, *, host: Optional[str] = None, port: int)
         if not host:
             host = socket.getfqdn()
 
+        if not ip:
+            network_binding = self.charm.model.get_binding(self.relation)
+            if network_binding is None:
+                raise RuntimeError(
+                    f"can't find IP address for relation {self.relation}, network binding is None"
+                )
+            binding_ip = network_binding.network.bind_address
+            if binding_ip is None:
+                raise RuntimeError(
+                    f"can't find IP address for relation {self.relation}, bind address is None"
+                )
+            ip = str(binding_ip)
+
```

Comment:
This raise is a bit dangerous as it's quite common for sidecar charms to not yet have a bind address during the first events executed. This could lead to the charm getting stuck in an error state.

Reply:
How about use `socket.gethostbyname(socket.gethostname())` only when the value isn't in the `127/8` range. This will likely decrease the risk of errors, though it won't entirely eliminate them.

Reply:
I think even in the case of being served a local loopback, we should go with that. Once it has access to a bind address, this would auto-heal, which I much prefer to it being **completely** broken in the meantime.

Reply:
> I've considered it, but I think the issue is primarily with Juju. When the charm responds to an event, the network should already be ready. And I am not sure if we should workaround that in the charm level.

Well, it would be a breaking change to current users as their ingress would potentially stop working even though they might not even use the ip (which is the case with traefik itself). Unfortunately, that will be the case regardless of whos "fault" it actually is. 😅

Reply:
I'm concerned about scenarios involving ingress controllers like Traffic or NGINX serving multiple applications. I'm not completely certain about the precise implementation, but if the ingress interprets the loopback address as-is, it could create a short circuit within the ingress. This might redirect traffic to a different application, potentially bypassing certain security measures.

For instance, if a requirer provide `127.0.0.1` as the endpoint address, it might generate a rule like the one below. This rule could route traffic from port `80` to endpoint `http://internal-app:8080`:

```
server {
  listen 8080;
  proxy_pass http://internal-app:8080;
}

server {
  listen 80;
  proxy_pass http://127.0.0.1:8080;
}
```

Reply:
> I'm concerned about scenarios involving ingress controllers like Traffic or NGINX serving multiple applications. I'm not completely certain about the precise implementation, but if the ingress interprets the loopback address as-is, it could create a short circuit within the ingress. This might redirect traffic to a different application, potentially bypassing certain security measures.
> 
> For instance, if a requirer provide `127.0.0.1` as the endpoint address, it might generate a rule like the one below. This rule could route traffic from port `80` to endpoint `http://internal-app:8080`:
> 
> ```
> server {
>   listen 8080;
>   proxy_pass http://internal-app:8080;
> }
> 
> server {
>   listen 80;
>   proxy_pass http://127.0.0.1:8080;
> }
> ```

Then I suggest we make it so that if it's getting a loopback (i.e `localhost`, `127.0.0.1` or `::1`, it instead sets it up to proxy pass to `about:blank`.

Reply:
Hopefully the `ip` field can be merged with the `host` (see comment above) and then:
- If we keep using fqdn as default then this issue goes away
- If we start using bind_address as a fallback for invalid fqdn then need a custom exception and add a `Raises:` section to the docstring.

> it's quite common for sidecar charms to not yet have a bind address during the first events executed

Seems like asking the caller to guard for these Nones would be more messy than try-except.

Reply:
> Seems like asking the caller to guard for these Nones would be more messy than try-except.

Well, the problem is going to be that it might have the charm go into a crash loop if the information isn't available when it checks. I at least can't for certain say that Juju will proceed with assigning an IP if the charm is stuck in `error`.

Reply:
Yeah it's an avenue we'd need to explore, but I hope we won't get there as I imagine we can have both merged into one arg, and always use fqdn as the only default. IP would have to be via kwarg.

Reply:
+1 to what leon said. Don't like the idea of adding it as a separate field to ingress.
We can consider wrapping that try/except in the library call, so the user doesn't have to remember doing it manually?

Reply:
Unfortunately, we cannot provide only the FQDN without the IP. This is because the Nginx Ingress Controller requires IP addresses as endpoints, not hosts. If the IP address is absent from the relation data, the `nginx-ingress-integrator` charm won't receive notifications of IP address changes. Therefore, to ensure compatibility with Nginx ingress in Kubernetes, it's essential to have the IP address present in the relation data.

As a solution, we could initially source the binding address from Juju. If that doesn't work We can then use `socket.gethostbyname(socket.gethostname())` if it's not in `127/8`. If all else fails, we can fallback to `socket.gethostbyname(socket.getfqdn())`. 

This approach ensures our library will never fail in environments in the environment mandated by Traefik-compatible applications (i.e. cluster DNS is enabled). And it will only fail when there isn't a cluster DNS, in that case, the Traefik charm can't function properly as well.

Reply:
How does the `nginx-ingress-integrator` charm run the `juju status` command?

When IP addresses are changed, the `nginx-ingress-integrator` needs to respond by updating the Kubernetes Nginx Ingress controller with the new IP address. This integrator operates based on Juju events. How can the `nginx-ingress-integrator` detect changes in IP addresses when they aren't present in the relation?

Reply:
I found some context from the previous meeting about the decision: https://chat.charmhub.io/charmhub/pl/m6mb68ijxjrwpkthsgf6xxju7r

Reply:
@PietroPasotti and @sed-i:

The issue here is not whether we should use ip or not in the traefik charm, cause we won't - we really don't have any reason to. The issue is rather that `ingress`, as a charm relation interface, needs to work with both nginx and traefik.

We could make ip optional. We could make host optional. The problem is that they'll be required in one interface implementation each. And if we make both required, the charms will crash when there is no bound ip yet, which is something traefik currently survives.

Reply:
ah so it's really a matter of making ingress interoperable between traefik and nginx?

Reply:
if so then I feel this is a discussion for canonical/charm-relation-interfaces

Reply:
> How does the `nginx-ingress-integrator` charm run the `juju status` command?
> 
> When IP addresses are changed, the `nginx-ingress-integrator` needs to respond by updating the Kubernetes Nginx Ingress controller with the new IP address. This integrator operates based on Juju events. How can the `nginx-ingress-integrator` detect changes in IP addresses when they aren't present in the relation?

It doesn't need to run `juju status` - that information is already in relation data:
```
$ juju show-unit am/0
am/0:
  opened-ports: []
  charm: ch:amd64/focal/alertmanager-k8s-88
  leader: true
  life: alive
  relation-info:
  - relation-id: 1
    endpoint: replicas
    related-endpoint: replicas
    application-data: {}
    local-unit:
      in-scope: true
      data:
        egress-subnets: 10.152.183.216/32
        ingress-address: 10.152.183.216
        private-address: 10.152.183.216
  - relation-id: 2
    endpoint: ingress
    related-endpoint: ingress
    application-data:
      ingress: '{"url": "http://10.43.8.206/test-am"}'
    related-units:
      trfk/0:
        in-scope: true
        data:
          egress-subnets: 10.43.8.206/32
          ingress-address: 10.43.8.206
          private-address: 10.43.8.206
  provider-id: am-0
  address: 10.1.166.95
```

Note how `private-address` is present for all relations' unit data.

Reply:
In Juju for Kubernetes, the `private-address` does not point to the unit's IP address; instead, it points to the cluster IP of service created by Juju for the entire application. To utilize that service, applications must use the `open-port`, which is only available in Juju 3.

Reply:
Shouldn't we organize a call to discuss and take a decision on that topic ? @sed-i @simskij @PietroPasotti

Reply:
@sed-i @simskij and @PietroPasotti Weii adapted based on our discussion yesterday, could you review it again ?